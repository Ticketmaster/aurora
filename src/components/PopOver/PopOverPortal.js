import { Component } from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";

class Portal extends Component {
  componentDidMount() {
    if (global.window.document.body) {
      this.rootEl = global.window.document.createElement("div");
      this.rootEl.id = `popover-root-${new Date().getTime()}`;
      global.window.document.body.appendChild(this.rootEl);
    }
  }

  componentWillUnmount() {
    global.window.document.body.removeChild(this.rootEl);
  }

  rootEl = null;

  render() {
    const { children } = this.props;

    return this.rootEl ? ReactDOM.createPortal(children, this.rootEl) : null;
  }
}

Portal.propTypes = {
  children: PropTypes.node
};

Portal.defaultProps = {
  children: null
};

export default Portal;
