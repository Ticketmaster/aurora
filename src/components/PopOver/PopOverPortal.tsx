import * as React from "react";
import * as PropTypes from "prop-types";
import { createPortal } from "react-dom";

const getRootNode = (node: HTMLElement): Node => {
  if (typeof node.getRootNode === "function") {
    const rootNode = node.getRootNode();

    if (rootNode && rootNode !== document) {
      return rootNode;
    }
  }

  return document.body;
};

class Portal extends React.Component<{}, { current?: HTMLElement }> {
  static propTypes = {
    children: PropTypes.node
  };

  static defaultProps = {
    children: null
  };

  state: { current?: HTMLElement } = {};

  componentWillUnmount(): void {
    const { current } = this.state;

    if (current && current.parentNode) {
      current.parentNode.removeChild(current);
    }
  }

  private refCallback = (ref: HTMLDivElement | null) => {
    if (ref) {
      const rootNode = getRootNode(ref);
      const current = document.createElement("div");

      current.id = `popover-root-${new Date().getTime()}`;

      rootNode.appendChild(current);

      this.setState({ current });
    }
  };

  render(): React.ReactNode {
    const { children } = this.props;
    const { current } = this.state;

    if (current) {
      return createPortal(children, current);
    }

    return <div ref={this.refCallback} style={{ display: "none" }} />;
  }
}

export default Portal;
