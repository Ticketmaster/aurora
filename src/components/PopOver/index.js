import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { CSSTransition } from "react-transition-group";
import { themes, constants, spacing, zIndex } from "../../theme";
import {
  popContainersBoxShadow,
  popContainersSharpBoxShadow
} from "../../theme/constants";
import PopOverPortal from "./PopOverPortal";

const StyledPopOver = styled.div`
  background-color: ${themes.global.white.base};
  ${({ noBorders }) =>
    !noBorders && `border: 1px solid ${themes.global.gray02}`};
  border-radius: ${constants.borderRadius.large};
  box-shadow: ${({ noBorders }) =>
    noBorders ? popContainersSharpBoxShadow : popContainersBoxShadow};
  position: absolute;
  max-width: 300px;
  padding: ${spacing.moderate};
  z-index: ${({ zInd }) => zInd};
  display: ${({ isVisible }) => (isVisible ? "block" : "none")};
  transition: opacity 0.1s ${constants.easing.easeInQuad},
    transform 0.1s ${constants.easing.easeInQuad};

  &.open-enter,
  &.open-appear {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad};
    display: block;
    opacity: 0;
    transform: scale(0.7);
  }

  &.open-enter-active,
  &.open-appear-active {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad},
      transform 0.3s ${constants.easing.easeInOutQuad};
    display: block;
    opacity: 1;
    transform: scale(1);
  }

  &.open-enter-done {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad},
      transform 0.3s ${constants.easing.easeInOutQuad};
    display: block;
    opacity: 1;
    transform: scale(1);
  }

  &.open-exit {
    display: block;
    opacity: 1;
    transform: scale(1);
  }

  &.open-exit-active {
    display: block;
    opacity: 0;
    transform: scale(0.7);
  }
`;

const MIN_SPACE_FROM_EDGE = 24;
const MOBILE_MIN_SPACE_FROM_EDGE = 16;
const MOBILE_MAX_WIDTH = 767;

class PopOver extends Component {
  /*
   * Static function that determines PopOver position.
   * @position(object) - top and bottom position of the elemt that triggers showing PopOver;
   * Mouse horizontal position on the sreen - so we can center the PopOver;
   * Additional container position and size
   * @dimensions(object) - PopOver width and height;
   * Page scroll position
   * Viewport size
   * @reduce(object) - additional top/bottom screen reduction cause by sticky header/footer
   * @inlineWithTarget(boolean) - an optional prop that allows to align PopOver with toggle
   */
  static calculatePosition({
    position,
    dimensions,
    reduce,
    inlineWithTarget,
    spaceFromMouse
  }) {
    const {
      width,
      windowScroll,
      height,
      windowWidth,
      windowHeight
    } = dimensions;
    const { top: reduceTop, bottom: reduceBottom } = reduce;
    const {
      elBottom,
      elTop,
      elLeft,
      elWidth,
      offsetTop,
      clientHeight,
      offsetLeft,
      clientWidth
    } = position;

    if (inlineWithTarget) {
      return {
        x:
          elLeft + width + MIN_SPACE_FROM_EDGE * 2 >= clientWidth
            ? elLeft - width - spacing.gutters.tiny
            : elLeft + elWidth + spacing.gutters.tiny,
        y: windowScroll + elTop
      };
    }

    const viewportTop = windowScroll + reduceTop;
    const viewportBottom = windowScroll + windowHeight - reduceBottom;
    const bottomPosition = windowScroll + elBottom + spaceFromMouse;
    const topPosition = windowScroll + elTop - spaceFromMouse - height;

    const spaceFromEdge =
      windowWidth > MOBILE_MAX_WIDTH
        ? MIN_SPACE_FROM_EDGE
        : MOBILE_MIN_SPACE_FROM_EDGE;

    const containerTop = offsetTop + spaceFromEdge;
    const containerBottom =
      windowScroll + offsetTop + clientHeight - spaceFromEdge;
    const containerLeft = offsetLeft + spaceFromEdge;
    const containerRight = offsetLeft + clientWidth - spaceFromEdge - width;

    const topPositionWithFallback =
      topPosition - spaceFromEdge >= Math.max(viewportTop, containerTop)
        ? topPosition
        : bottomPosition;

    return {
      x: Math.min(
        Math.max(elLeft + elWidth / 2 - width / 2, containerLeft),
        windowWidth - spaceFromEdge - width,
        containerRight
      ),
      y:
        bottomPosition + height + spaceFromEdge <=
        Math.min(viewportBottom, containerBottom)
          ? bottomPosition
          : topPositionWithFallback
    };
  }

  constructor(props) {
    super(props);

    this.dimensions = {
      width: 0,
      height: 0,
      windowScroll: 0,
      windowWidth: 0,
      windowHeight: 0
    };

    this.myRef = React.createRef();

    this.pos = {
      x: 0,
      y: 0
    };
  }

  componentDidUpdate(prevProps) {
    /*
     * This causes force update of the popover position if we use only one popover and we transiotion from
     * one zone that triggers the popover to another that triggers the same popover but with different place to display
     */
    const {
      position: { elTop, elBottom },
      isVisible,
      inlineWithTarget,
      position,
      reduceTop,
      reduceBottom,
      spaceFromMouse
    } = this.props;

    if (
      (prevProps.position.elTop !== elTop ||
        prevProps.position.elBottom !== elBottom) &&
      isVisible &&
      prevProps.isVisible === isVisible
    ) {
      this.setDimensions();

      const reduce = { top: reduceTop, bottom: reduceBottom };
      this.pos = PopOver.calculatePosition({
        position,
        dimensions: this.dimensions,
        reduce,
        inlineWithTarget,
        spaceFromMouse
      });

      this.myRef.current.style.top = `${this.pos.y}px`;
      this.myRef.current.style.left = `${this.pos.x}px`;
    }
  }

  /*
   * Static function that needs to be called from the parent -> PopOver.getDimensionsFromEvent
   * The parent should pass the click event which will trigger showing the PopOver.
   * By default the PopOver is shown withing the view port. If we need to show it inside
   * certain element we should pass the html element as second parameter.
   * The function will return an object that should be provided to the PopOver as props.
   */
  static getDimensionsFromEvent(e, parent) {
    const {
      y: elTop,
      height: elHeight,
      x: elLeft,
      width: elWidth
    } = e.target.getBoundingClientRect();
    const {
      y: offsetTop = 0,
      height: clientHeight = 100000,
      x: offsetLeft = 0,
      width: clientWidth = 100000
    } =
      parent && parent.getBoundingClientRect
        ? parent.getBoundingClientRect()
        : {};

    return {
      elTop,
      elLeft,
      elWidth,
      elBottom: elTop + elHeight,
      offsetTop,
      clientHeight,
      offsetLeft,
      clientWidth
    };
  }

  setDimensions = () => {
    const { isVisible } = this.props;
    const {
      windowScroll,
      windowWidth,
      windowHeight,
      width,
      height
    } = this.dimensions;
    const dimensions = {};
    if (global.window && isVisible) {
      const {
        scrollTop,
        clientWidth,
        clientHeight
      } = global.window.document.documentElement;

      if (scrollTop !== windowScroll) {
        dimensions.windowScroll = scrollTop;
      }

      if (clientWidth !== windowWidth) {
        dimensions.windowWidth = clientWidth;
      }

      if (clientHeight !== windowHeight) {
        dimensions.windowHeight = clientHeight;
      }
    }

    if (this.myRef.current) {
      const { clientWidth, clientHeight } = this.myRef.current;

      if (width !== clientWidth && clientWidth) {
        dimensions.width = clientWidth;
      }

      if (height !== clientHeight && clientHeight) {
        dimensions.height = clientHeight;
      }
    }

    if (Object.keys(dimensions).length) {
      this.dimensions = {
        ...this.dimensions,
        ...dimensions
      };
      return true;
    }

    return false;
  };

  popoverEnter = () => {
    const {
      isVisible,
      inlineWithTarget,
      position,
      reduceTop,
      reduceBottom,
      spaceFromMouse
    } = this.props;

    if (isVisible) {
      this.setDimensions();

      const reduce = { top: reduceTop, bottom: reduceBottom };
      this.pos = PopOver.calculatePosition({
        position,
        dimensions: this.dimensions,
        reduce,
        inlineWithTarget,
        spaceFromMouse
      });
    }

    this.myRef.current.style.top = `${this.pos.y}px`;
    this.myRef.current.style.left = `${this.pos.x}px`;
  };

  render() {
    const { children, isVisible, noBorders, zInd } = this.props;

    return (
      <PopOverPortal>
        <CSSTransition
          in={isVisible}
          key="popover-animation"
          timeout={300}
          classNames="open"
          appear={isVisible}
          onEnter={this.popoverEnter}
        >
          <StyledPopOver
            ref={this.myRef}
            isVisible={isVisible}
            noBorders={noBorders}
            zInd={zInd}
          >
            {children}
          </StyledPopOver>
        </CSSTransition>
      </PopOverPortal>
    );
  }
}

PopOver.propTypes = {
  children: PropTypes.node.isRequired,
  isVisible: PropTypes.bool,
  inlineWithTarget: PropTypes.bool,
  noBorders: PropTypes.bool,
  reduceTop: PropTypes.number,
  reduceBottom: PropTypes.number,
  position: PropTypes.shape({
    mouseX: PropTypes.number,
    elTop: PropTypes.number,
    elBottom: PropTypes.number,
    offsetTop: PropTypes.number,
    clientHeight: PropTypes.number,
    offsetLeft: PropTypes.number,
    clientWidth: PropTypes.number
  }),
  zInd: PropTypes.number,
  spaceFromMouse: PropTypes.number
};

PopOver.defaultProps = {
  isVisible: false,
  inlineWithTarget: false,
  noBorders: false,
  reduceTop: 0,
  reduceBottom: 0,
  position: {
    mouseX: 0,
    elTop: 0,
    elBottom: 0,
    offsetTop: 0,
    clientHeight: 0,
    offsetLeft: 0,
    clientWidth: 0
  },
  zInd: zIndex.layout.overlay,
  spaceFromMouse: 4
};

PopOver.displayName = "PopOver";

export default PopOver;
