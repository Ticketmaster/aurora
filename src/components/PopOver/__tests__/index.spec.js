import React from "react";
import renderer from "react-test-renderer";

import PopOver from "..";

jest.mock("../PopOverPortal", () => ({ children }) => children);

jest.mock("react-transition-group", () => ({
  CSSTransition: ({ children }) => children
}));

describe("PopOver", () => {
  it("should match snapshot", () => {
    const tree = renderer.create(<PopOver />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should match snapshot when visible", () => {
    const tree = renderer.create(<PopOver isVisible />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should match snapshot when visible and without borders", () => {
    const tree = renderer.create(<PopOver isVisible noBorders />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe("calculatePosition tests", () => {
    const reduce = {
      top: 0,
      bottom: 0
    };
    it("should render viewport left and bottom", () => {
      const position = {
        elBottom: 10,
        elTop: 0,
        offsetTop: 0,
        clientHeight: 1000,
        offsetLeft: 0,
        clientWidth: 500,
        elWidth: 10,
        elLeft: 0
      };
      const dimensions = {
        width: 100,
        windowScroll: 0,
        height: 100,
        windowWidth: 500,
        windowHeight: 400
      };

      const spaceFromMouse = 4;

      expect(
        PopOver.calculatePosition({
          position,
          dimensions,
          reduce,
          spaceFromMouse
        })
      ).toEqual({
        x: 16,
        y: 14
      });
    });

    it("should render viewport left and bottom on wide screen", () => {
      const position = {
        elBottom: 510,
        elTop: 500,
        elLeft: 0,
        elWidth: 10,
        offsetTop: 0,
        clientHeight: 1000,
        offsetLeft: 0,
        clientWidth: 10000
      };

      const dimensions = {
        width: 100,
        windowScroll: 100,
        height: 100,
        windowWidth: 1000,
        windowHeight: 400
      };
      const spaceFromMouse = 4;

      expect(
        PopOver.calculatePosition({
          position,
          dimensions,
          reduce,
          spaceFromMouse
        })
      ).toEqual({
        x: 24,
        y: 496
      });
    });

    it("should position PopOver near toggle to the right", () => {
      const position = {
        elBottom: 510,
        elTop: 500,
        elLeft: 30,
        elWidth: 150,
        offsetTop: 0,
        clientHeight: 1000,
        offsetLeft: 0,
        clientWidth: 1000
      };
      const dimensions = {
        width: 100,
        windowScroll: 100,
        height: 100,
        windowWidth: 1000,
        windowHeight: 400
      };
      const spaceFromMouse = 4;

      expect(
        PopOver.calculatePosition({
          position,
          dimensions,
          reduce,
          inlineWithTarget: true,
          spaceFromMouse
        })
      ).toEqual({
        x: 188,
        y: 600
      });
    });

    it("should position PopOver near toggle to the left", () => {
      const position = {
        elBottom: 510,
        elTop: 500,
        elLeft: 900,
        elWidth: 150,
        offsetTop: 0,
        clientHeight: 1000,
        offsetLeft: 0,
        clientWidth: 1000
      };
      const dimensions = {
        width: 100,
        windowScroll: 100,
        height: 100,
        windowWidth: 1000,
        windowHeight: 400
      };
      const spaceFromMouse = 4;

      expect(
        PopOver.calculatePosition({
          position,
          dimensions,
          reduce,
          inlineWithTarget: true,
          spaceFromMouse
        })
      ).toEqual({
        x: 792,
        y: 600
      });
    });
  });

  it("popoverEnter should calculate position again and set it", () => {
    const setDimensionsMock = jest.fn();
    const calculatePositionMock = jest.fn(() => ({ x: 5, y: 10 }));
    const tree = renderer.create(<PopOver isVisible />).getInstance();

    tree.setDimensions = setDimensionsMock;
    PopOver.calculatePosition = calculatePositionMock;
    tree.myRef = {
      current: {
        style: {
          top: 0,
          left: 0
        }
      }
    };
    tree.popoverEnter();

    expect(setDimensionsMock).toHaveBeenCalledTimes(1);
    expect(calculatePositionMock).toHaveBeenCalledTimes(1);
    expect(tree.myRef.current.style.top).toEqual("10px");
    expect(tree.myRef.current.style.left).toEqual("5px");
  });

  it("popoverEnter should not calculate position again and set old pos values", () => {
    const setDimensionsMock = jest.fn();
    const calculatePositionMock = jest.fn(() => ({ x: 5, y: 10 }));
    const tree = renderer.create(<PopOver />).getInstance();

    tree.setDimensions = setDimensionsMock;
    PopOver.calculatePosition = calculatePositionMock;
    tree.myRef = {
      current: {
        style: {
          top: 0,
          left: 0
        }
      }
    };
    tree.pos = {
      x: 5,
      y: 10
    };
    tree.popoverEnter();

    expect(setDimensionsMock).toHaveBeenCalledTimes(0);
    expect(calculatePositionMock).toHaveBeenCalledTimes(0);
    expect(tree.myRef.current.style.top).toEqual("10px");
    expect(tree.myRef.current.style.left).toEqual("5px");
  });

  it("componentDidUpdate should calculate position again and set it", () => {
    const setDimensionsMock = jest.fn();
    const calculatePositionMock = jest.fn(() => ({ x: 5, y: 10 }));
    const position = {
      elTop: 20,
      elBottom: 40
    };
    const tree = renderer.create(<PopOver isVisible />).getInstance();

    tree.setDimensions = setDimensionsMock;
    PopOver.calculatePosition = calculatePositionMock;
    tree.myRef = {
      current: {
        style: {
          top: 0,
          left: 0
        }
      }
    };
    tree.componentDidUpdate({ position, isVisible: true });

    expect(setDimensionsMock).toHaveBeenCalledTimes(1);
    expect(calculatePositionMock).toHaveBeenCalledTimes(1);
    expect(tree.myRef.current.style.top).toEqual("10px");
    expect(tree.myRef.current.style.left).toEqual("5px");
  });

  it("componentDidUpdate should not do anything if not visible", () => {
    const setDimensionsMock = jest.fn();
    const calculatePositionMock = jest.fn(() => ({ x: 5, y: 10 }));
    const position = {
      elTop: 20,
      elBottom: 40
    };
    const tree = renderer
      .create(<PopOver isVisible position={position} />)
      .getInstance();

    tree.setDimensions = setDimensionsMock;
    PopOver.calculatePosition = calculatePositionMock;
    tree.myRef = {
      current: {
        style: {
          top: 0,
          left: 0
        }
      }
    };
    tree.componentDidUpdate({ position });

    expect(setDimensionsMock).toHaveBeenCalledTimes(0);
    expect(calculatePositionMock).toHaveBeenCalledTimes(0);
    expect(tree.myRef.current.style.top).toEqual(0);
    expect(tree.myRef.current.style.left).toEqual(0);
  });

  it("setDimensions should not call setState and return false", () => {
    const tree = renderer.create(<PopOver isVisible />).getInstance();

    tree.setDimensions();

    expect(tree.setDimensions()).toBeFalsy();
  });

  it("setDimensions should not call setState and return false when pop over dimensions are 0, 0 pop over isn ot visible", () => {
    const tree = renderer.create(<PopOver />).getInstance();

    tree.dimensions.width = 100;
    tree.dimensions.height = 100;
    tree.myRef = {
      current: {
        clientWidth: 0,
        clientHeight: 0
      }
    };
    tree.setDimensions();

    expect(tree.setDimensions()).toBeFalsy();
  });

  describe("getDimensionsFromEvent", () => {
    const event = {
      target: {
        getBoundingClientRect: () => ({
          y: 20,
          height: 1000,
          x: 30,
          width: 150
        })
      }
    };
    const wrapper = {
      getBoundingClientRect: () => ({
        y: 50,
        height: 200,
        x: 100,
        width: 300
      })
    };
    it("without wrapper", () => {
      expect(PopOver.getDimensionsFromEvent(event)).toEqual({
        elTop: 20,
        elLeft: 30,
        elWidth: 150,
        elBottom: 1020,
        offsetTop: 0,
        clientHeight: 100000,
        offsetLeft: 0,
        clientWidth: 100000
      });
    });

    it("plus wrapper", () => {
      expect(PopOver.getDimensionsFromEvent(event, wrapper)).toEqual({
        elTop: 20,
        elLeft: 30,
        elWidth: 150,
        elBottom: 1020,
        offsetTop: 50,
        clientHeight: 200,
        offsetLeft: 100,
        clientWidth: 300
      });
    });
  });

  describe("when define global window", () => {
    it("setDimensions should call setState and return true", () => {
      Object.defineProperty(
        global.window.document.documentElement,
        "clientHeight",
        {
          value: 1000,
          writable: true
        }
      );
      Object.defineProperty(
        global.window.document.documentElement,
        "clientWidth",
        {
          value: 1000,
          writable: true
        }
      );
      Object.defineProperty(
        global.window.document.documentElement,
        "scrollTop",
        {
          value: 100,
          writable: true
        }
      );
      const tree = renderer.create(<PopOver isVisible />).getInstance();

      tree.myRef = {
        current: {
          clientWidth: 200,
          clientHeight: 200
        }
      };

      expect(tree.setDimensions()).toBeTruthy();
      expect(tree.dimensions).toEqual({
        windowScroll: 100,
        windowWidth: 1000,
        windowHeight: 1000,
        width: 200,
        height: 200
      });
    });
  });
});
