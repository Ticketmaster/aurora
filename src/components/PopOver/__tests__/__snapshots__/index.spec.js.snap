// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopOver should match snapshot 1`] = `
.c0 {
  background-color: rgba(255,255,255,1);
  border: 1px solid #999999;
  border-radius: 4px;
  box-shadow: 0 4px 4px 0 rgba(0,0,0,0.12);
  position: absolute;
  max-width: 260px;
  padding: 16px;
  display: none;
  -webkit-transition: opacity 0.1s cubic-bezier(0.55,0.085,0.68,0.53), -webkit-transform 0.1s cubic-bezier(0.55,0.085,0.68,0.53);
  -webkit-transition: opacity 0.1s cubic-bezier(0.55,0.085,0.68,0.53), transform 0.1s cubic-bezier(0.55,0.085,0.68,0.53);
  transition: opacity 0.1s cubic-bezier(0.55,0.085,0.68,0.53), transform 0.1s cubic-bezier(0.55,0.085,0.68,0.53);
}

.c0.open-enter {
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  display: block;
  opacity: 0;
  -webkit-transform: scale(0.7);
  -ms-transform: scale(0.7);
  transform: scale(0.7);
}

.c0.open-enter-active {
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), -webkit-transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  display: block;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c0.open-enter-done {
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), -webkit-transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  display: block;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c0.open-exit {
  display: block;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c0.open-exit-active {
  display: block;
  opacity: 0;
  -webkit-transform: scale(0.7);
  -ms-transform: scale(0.7);
  transform: scale(0.7);
}

<div
  className="c0"
/>
`;

exports[`PopOver should match snapshot 2`] = `
.c0 {
  background-color: rgba(255,255,255,1);
  border: 1px solid #999999;
  border-radius: 4px;
  box-shadow: 0 4px 4px 0 rgba(0,0,0,0.12);
  position: absolute;
  max-width: 260px;
  padding: 16px;
  display: block;
  -webkit-transition: opacity 0.1s cubic-bezier(0.55,0.085,0.68,0.53), -webkit-transform 0.1s cubic-bezier(0.55,0.085,0.68,0.53);
  -webkit-transition: opacity 0.1s cubic-bezier(0.55,0.085,0.68,0.53), transform 0.1s cubic-bezier(0.55,0.085,0.68,0.53);
  transition: opacity 0.1s cubic-bezier(0.55,0.085,0.68,0.53), transform 0.1s cubic-bezier(0.55,0.085,0.68,0.53);
}

.c0.open-enter {
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  display: block;
  opacity: 0;
  -webkit-transform: scale(0.7);
  -ms-transform: scale(0.7);
  transform: scale(0.7);
}

.c0.open-enter-active {
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), -webkit-transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  display: block;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c0.open-enter-done {
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), -webkit-transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  -webkit-transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  transition: opacity 0.3s cubic-bezier(0.455,0.03,0.515,0.955), transform 0.3s cubic-bezier(0.455,0.03,0.515,0.955);
  display: block;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c0.open-exit {
  display: block;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c0.open-exit-active {
  display: block;
  opacity: 0;
  -webkit-transform: scale(0.7);
  -ms-transform: scale(0.7);
  transform: scale(0.7);
}

<div
  className="c0"
/>
`;
